bpo-38707: Fix for multiprocessing.Process MainThread.native_id

jaketesler opened the pull request:
This PR implements a fix for `multiprocessing.Process` objects; the error occurs when Processes are created using either `fork` or `forkserver` as the `start_method`.

In these instances, the `MainThread` of the newly created `Process` object retains all attributes from its parent's `MainThread` object, including the `native_id` attribute. The resulting behavior is such that the new process' `MainThread` captures an incorrect/outdated `native_id` (the parent's instead of its own). 

This change forces the Process object to update its `native_id` attribute during the bootstrap process.

cc @vstinner

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-38707](https://bugs.python.org/issue38707) -->
https://bugs.python.org/issue38707
<!-- /issue-number -->


Automerge-Triggered-By: @pitrou

jaketesler committed:
Implement fix for MainThread.native_id for multiprocessing.Process objects created via fork start method

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

brandtbucher commented on Misc/NEWS.d/next/Core and Builtins/2019-11-08-00-36-10.bpo-38707.SZL036.rst:
```diff
@@ -0,0 +1 @@
+`multiprocessing.Process` objects now retain their correct `MainThread.native_id` attribute
```
Just a bit of markup here:

```suggestion
:class:`multiprocessing.Process` objects now retain their correct ``MainThread.native_id`` attribute.
```

jaketesler replied:
A test for this would be relatively simple – I’ll write one and add it to the PR. Good suggestion!

brandtbucher reviewed (changes requested):
Not a multiprocessing expert, but this seems clean enough! How hard would it be to add a test for this behavior?

jaketesler committed:
Update Misc/NEWS.d/next/Core and Builtins/2019-11-08-00-36-10.bpo-38707.SZL036.rst

Co-Authored-By: Brandt Bucher <brandtbucher@gmail.com>

jaketesler committed:
Add multiprocessing native_id test

jaketesler committed:
Skip test if type is 'threads'

Since Dummy Processes are simply Thread objects,
they will share a MainThread and their native_id attribute will match.

jaketesler committed:
Minor code cleanup

brandtbucher commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -361,6 +361,27 @@ def test_process(self):
         self.assertNotIn(p, self.active_children())
         close_queue(q)
 
+    if threading._HAVE_THREAD_NATIVE_ID:
```
I think it's more conventional here to use `@unittest.skipUnless(threading._HAVE_THREAD_NATIVE_ID, "needs native_id")` instead of this branching. Then both of these methods can be unindented.

`_test_process_thread_attributes` can also be a `@staticmethod` (`cls` isn't used). And personally, I would recommend renaming it to something less test-like and more unique... but that's just me!

jaketesler replied:
On the first, I was following the convention in the `threading` module but this seems more reasonable. 

While `_test_process_thread_attributes` could be a `staticmethod`, I would prefer to keep consistency within this class – the other target functions all use `@classmethod` regardless if they use the `cls` variable or not. 
I should probably rename the function though, good point.

jaketesler committed:
Refactoring per suggestions from @brandtbucher

brandtbucher reviewed (approved):
Looks good!

pitrou commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -361,6 +361,28 @@ def test_process(self):
+
+        q = self.Queue(1)
+        p = self.Process(target=self._test_process_mainthread_native_id, args=(q,))
+        p.daemon = True
```
Rather than making it a daemon process, it's better to call `p.join()` below.

pitrou reviewed (changes requested):
LGTM except for one minor comment.

jaketesler committed:
Modified tests' child Process from daemon to using join() per @pitrou

pitrou committed:
Improve NEWS message

pitrou reviewed (approved):
+1
